tools/windows/install-help/cal/PropertyReplacer.cpp

```diff
@@ -0,0 +1,180 @@
+#include "stdafx.h"
+#include "PropertyReplacer.h"
+
+namespace
+{
+    template <class Map>
+    bool has_key(Map const &m, const typename Map::key_type &key)
+    {
+        auto const &it = m.find(key);
+        return it != m.end();
+    }
+
+    typedef std::function<std::wstring(std::wstring const &, const property_retriever &)> formatter_func;
+
+    /// <summary>
+    /// Simply concatenates <paramref name="str"/> with the value of the matching property.
+    /// </summary>
+    /// <param name="str">The string to use as a replacement.</param>
+    /// <returns>A function that conforms to <see cref="formatter_func"> that when called with a property value,
+    /// will return a concatenated string of <paramref name="str"/> and the property value. </returns>
+    formatter_func format_simple_value(const std::wstring &str)
+    {
+        return [str](std::wstring const &propertyValue, const property_retriever &)
+        {
+            return str + propertyValue;
+        };
+    }
+
+    std::wstring format_tags(const std::wstring &tags, const property_retriever &)
+    {
+        std::wistringstream valueStream(tags);
+        std::wstringstream result;
+        std::wstring token;
+        result << L"tags: ";
+        while (std::getline(valueStream, token, wchar_t(',')))
+        {
+            result << std::endl << L"  - " << token;
+        }
+        return result.str();
+    };
+
+    std::wstring format_proxy(const std::wstring &proxyHost, const property_retriever &propertyRetriever)
+    {
+        const auto &proxyPort = propertyRetriever(L"PROXY_PORT");
+        const auto &proxyUser = propertyRetriever(L"PROXY_USER");
+        const auto &proxyPassword = propertyRetriever(L"PROXY_PASSWORD");
```

Comment:
These return values are temporary `std::optional` objects crated inside the `propertyRetriever` lambda, don't receive them as references.